### ISSUE TITLE
File system crashes when S3 provider returns HTTP 429

### ISSUE THREAD

**Issue by kescherCode** [s3ql/s3ql#276]
Object storage provider is https://contabo.com.

When doing a lot with the object storage, for instance when running an `s3qlrm` command on a large file, it will occasionally return HTTP 429. The entire s3ql filesystem crashes when this happens, and can only be unmounted using `fusermount -u`, not `umount.s3ql`.

```
Traceback (most recent call last):
File "/usr/lib/python3.10/site-packages/s3ql/mount.py", line 58, in run_with_except_hook
  run_old(*args, **kw)
File "/usr/lib/python3.10/threading.py", line 946, in run
  self._target(*self._args, **self._kwargs)
File "/usr/lib/python3.10/site-packages/s3ql/block_cache.py", line 650, in _removal_loop_simple
  backend.delete('s3ql_data_%d' % id_)
File "/usr/lib/python3.10/site-packages/s3ql/backends/comprenc.py", line 289, in delete
  return self.backend.delete(key, force)
File "/usr/lib/python3.10/site-packages/s3ql/backends/common.py", line 108, in wrapped
  return method(*a, **kw)
File "/usr/lib/python3.10/site-packages/s3ql/backends/s3c.py", line 219, in delete
  resp = self._do_request('DELETE', '/%s%s' % (self.prefix, key))
File "/usr/lib/python3.10/site-packages/s3ql/backends/s3c.py", line 532, in _do_request
  self._parse_error_response(resp)
File "/usr/lib/python3.10/site-packages/s3ql/backends/s3c.py", line 554, in _parse_error_response
  raise HTTPError(resp.status, resp.reason, resp.headers)
s3ql.backends.s3c.HTTPError: 429 Too Many Requests
```

**Comment by kescherCode** [comment]
(I may take a look at this myself and might write a pull request at some point)

**Comment by kescherCode** [comment]
My suggested fix would be to do a "backoff" strategy - if 429 happens, just retry after an increasing amount of time. First start with a few seconds, and then get to minutes and eventually retry every 24 hours. Perhaps evaluate whether the same should be done for the metadata backup that's done every 24 hours.

**Comment by d--j** [comment]
Hello @kescherCode,
if you want to do the PR for this yourself, here's the place you need to add the 429 HTTP status code: https://github.com/s3ql/s3ql/blob/896e1debfb4886329e96bcef90860a3d37956094/src/s3ql/backends/s3c.py#L156-L160

The retry/backoff code already is in place. Metadata uploads also use this code path.

**Comment by kescherCode** [comment]
Thanks for the hint! I will probably get to it in a few hours.

**Comment by kescherCode** [comment]
Interesting, this had already been implemented for Swift... https://github.com/s3ql/s3ql/blob/896e1debfb4886329e96bcef90860a3d37956094/src/s3ql/backends/swift.py#L129-L134

and B2. https://github.com/s3ql/s3ql/blob/896e1debfb4886329e96bcef90860a3d37956094/src/s3ql/backends/b2/b2_backend.py#L372-L376

**Comment by kescherCode** [comment]
Fixed by #277
